docs/DesignDocs/SGX2-EDMM-driver-interface.md

```diff
@@ -0,0 +1,137 @@
+ */
+struct sgx_enclave_mprotect {
+	__u64	addr;
+	__u64	length;
+	__u64	prot;
+};
+```
+
+### Remarks
+
+Kernel should ensure that SGX instructions can succeed or catch and handle any fault.
+  - The kernel may maintain EPCM information on each page which includes access permission RWX, page types of PT_REG, PT_TRIM, PT_TCS.
+  - The kernel catches fault on EMODPR, EMODT and converts to error code returned to user space.
+
+The enclave run-time (or trusted run-time) may implement a parallel memory management structure which would provide information to the enclave on the enclave memory mappings.  The run-time can have a trusted API analogous to mmap which makes a call out of the enclave to issue the mmap  and then either perform EACCEPT on the pages and update the internal memory structures or configure the enclave to perform the EACCEPT when a #PF is delivered to the enclave.  With the outlined kernel interface, either implementation is possible.
+
+## Example advanced flows
+
+More advanced flows can be implemented as combinations of the basic flows. Here we present a few examples. 
+
+### Dynamic code loading
+
+An enclave can load trusted code to a new EPC page with execution permission using EACCEPT_COPY with following steps:
+1. Enclave calls mmap for a region in enclave ELRANGE for EAUG
+2. Enclave issues EACCEPT_COPY to copy trusted code from an existing EPC page to an unavailable page at an address inside that region
+   - Mechanism for enclave to establish the trustworthiness of new code is out-of-scope here.
+   - This results a #PF
+3. Kernel EAUG the page on #PF, set RW permissions in both EPCM and PTE
+4. The fault handler returns,  EACCEPT_COPY instruction retried, which sets RX permissions in EPCM specified in PageInfo operand.
+5. Enclave makes ocall which invokes mprotect syscall to change PTE permissions from RW to RX 
+
+### Lazy dynamic stack expansion
+An enclave can lazily expand its stacks as follows.
+1. Enclave calls mmap with MAP_GROWSDOWN for a stack region in enclave ELRANGE
+2. At some time later, enclave pushes to the top of the stack where no EPC page is populated yet, this results in #PF, causing enclave AEX.
+3. Kernel determines faulting address is in a stack region in enclave, EAUGs a page, injects a SIGBUS signal to enclave hosting process.
+4. User space untrusted side signal handler catches the SIGBUS and delivers it into enclave exception handler
+5. The enclave exception handler checks the faulting address against its record, determines the fault has happened in a stack area not yet EACCEPT'ed.
+6. Enclave issues EACCEPT, returns to untrusted signal handler of the hosting process, which returns to kernel fault handler.
+7. Kernel fault handler returns to enclave AEX address at which an ERESUME instruction is stored
+8. Enclave resumed and the original push instruction is retried and succeeds.
+
+
```

Comment:
There is one corner-case that is missing from this document (but mentioned in the original paper).

Typically there are only two SSAs in SGX enclaves: one for normal enclave execution and one for exception handling. But with on-demand enclave pages (where #PF is transformed into a SIGSEGV/SIGBUS signal and sent to the enclave runtime), there can be a legit case of `in-enclave exception during first-exception handling`. E.g., an enclave may be inside the SIGILL signal handler (to e.g. emulate `rdtsc`), and it will try to access a lazily allocated page, which will trigger a #PF/SIGBUS. But since we are already in SSA=1, all SSAs are occupied, and the enclave cannot make progress, and the whole application hangs (or killed by the SGX driver).

So it looks like enclave runtime (SGX SDK, OpenEnclave, Graphene, ...) must adopt a model with 3 SSAs, where one SSA is just-in-case to resolve PFs. I would definitely add this into this document, it's a tricky issue (especially in multi-threaded apps).

Reply:
@dimakuv, in your case of SIGILL, the handler (SSA=1) would **not** emulate `rdtsc` directly. Rather, it pushes `SSA[0].rip` and then points it to the real handler for `rdtsc`, then it exits. The real handler will run in SSA=0.

Reply:
Regarding #of SSAs needed.
In our SDK, there is so called two-phase exception handling in TRTS side which uses only two SSAs.
First phase is the initial signal handler entry, always use SSA [1], do minimal to setup stack and entry point for second phase, and set IP to the second phase entry (https://github.com/intel/linux-sgx/blob/544159a94a59547a18b7bff1d28a94bb14053572/sdk/trts/trts_veh.cpp#L431), then returns to untrusted side sighandler, which immediately ERESUME. That would transfer control flow to second phase entry with SSA[0].
All exceptions only happen with SSA[0]. And SSA[1] is used only in phase 1. Exception/faults can happen recursively and stacked up exception->phase1->phase2->exception->phase1->...

At the end of second phase exception handling, control transferred to IP that caused the last exception (https://github.com/intel/linux-sgx/blob/544159a94a59547a18b7bff1d28a94bb14053572/sdk/trts/trts_veh.cpp#L278)

Reply:
@binxing and @haitaohuang Thanks for reminding me this!

Yes, the same logic applies to Graphene. My bad, I forgot how exception handling works in SGX runtimes :)

So then there is no real issue. If the "SSA1" exception handler of the SGX runtime is correctly written, then it will **never** access a lazily allocated enclave page (in fact, I don't think the exception handler touches any memory at all except for SSA frames), then the issue I described in the previous comment does not exist.